content/en/BestPractices/prevent-xml-external-entities-attack.md

```diff
@@ -0,0 +1,292 @@
+
+- Read Internal Local Files 
+- Sensitive Data Exposure 
+- Server-Side Request Forgery 
+- Denial of Service 
+- Remote Code Execution 
+
+
+## Attack Scenarios
+
+Let’s see some vulnerable code examples to understand the XXE attacks further.
+
+
+### XXE in PHP Framework
+
+The below code example uses a PHP endpoint to parse the XML input and utilizes the `php-xml` module to perform the XML parsing. However, due to the support of external DTD parsing, an attacker could launch an XXE attack.
+
+```
+<?php
+
+require_once('../_helpers/strip.php');
+
+
+libxml_disable_entity_loader (false);
+
+$xml = strlen($_GET['xml']) > 0 ? $_GET['xml'] : '<root><content>No XML found</content></root>';
+
+$document = new DOMDocument();
+$document->loadXML($xml, LIBXML_NOENT | LIBXML_DTDLOAD);
+$parsedDocument = simplexml_import_dom($document);
+
+echo $parsedDocument->content;
+
+```
+
+An attacker can use the `expect://` function of the PHP to perform a remote code execution using XXE, and the payload will look like this: 
+
+```
+<?xml version="1.0" encoding="ISO-8859-1"?>
+<!DOCTYPE foo [ <!ELEMENT foo ANY >
+<!ENTITY xxe SYSTEM "expect://id" >]>
+<creds>
+    <user>&xxe;</user>
+    <pass>mypass</pass>
+</creds>
+```
+
+The command `expect://id` will return the output of the `id` command of UNIX, which typically produces output such as `uid=0(root) gid=0(root) groups=0(root)`. To remediate such scenarios, you need to disable the unnecessary protocols and functions such as `expect://` in the above example and disable external DTD parsing.
+
+
+### XXE in JavaScript 
+
+JavaScript framework such as `Node.Js` does not provide native XML parsing capabilities. To allow XML parsing, the `libxml` library can be used as shown in the code below: 
+
+```
+const app = require("express")(),
+const libxml = require("libxmljs");
+app.post("/profile/add", (req, res) => {
+  favorite = libxml.parseXml(req.body, { noent: true }); //noent is set to true
+  addToFavorites(favorite)
+});
+```
+
+In the above code, the `noent` property is set as `true`, allowing for external entity parsing leading to an XXE attack. An attacker may craft a malicious payload to make the server parse the external DTD and launch a successful XXE attack. 
+
+From the developer's perspective, avoid setting the `noent` property to `true`, which is disabled by default in the `libxmljs`. For example, the below code is an example of a secure code that does not allow external DTD parsing. 
+
+```
+const app = require("express")(),
+const libxml = require("libxmljs");
+app.post("/profile/add", (req, res) => {
+  favorite = libxml.parseXml(req.body); //noent is not set to true
+  addToFavorites(favorite)
+});
+
+```
+### XXE on Zend Framework via PHP FPM
+
+This code example is based on  CVE-2015-5161, which exploits a vulnerability in the Zend Framework leading to XXE attack. The following code is an illustration of a PHP application using Zend Framework to implement XML-RPC call:
+
+```
```

Comment:
```suggestion
```

### Zend Framework through PHP FPM

This code example is based on the Zend Framework vulnerability [CVE-2015-5161](https://nvd.nist.gov/vuln/detail/CVE-2015-5161). It illustrates how a PHP application uses the Zend Framework to implement an XML-RPC call:

```php
```